cpp/src/parquet/key_toolkit.h

```diff
@@ -0,0 +1,123 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "parquet/key_encryption_key.h"
+#include "parquet/kms_client.h"
+#include "parquet/kms_client_factory.h"
+#include "parquet/platform.h"
+#include "parquet/two_level_cache_with_expiration.h"
+
+namespace parquet {
+namespace encryption {
+
+class KeyWithMasterID {
+ public:
+  KeyWithMasterID(const std::string& key_bytes, const std::string& master_id)
+      : key_bytes_(key_bytes), master_id_(master_id) {}
+
+  const std::string& data_key() const { return key_bytes_; }
+  const std::string& master_id() const { return master_id_; }
+
+ private:
+  std::string key_bytes_;
+  std::string master_id_;
+};
+
+class PARQUET_EXPORT KeyToolkit {
+ public:
+  class KmsClientCache {
+   public:
+    static KmsClientCache& GetInstance() {
+      static KmsClientCache instance;
```

Comment:
could you discuss why a singleton is used here?  Is there never a case where two caches could be useful?

Reply:
to @ggershinsky

Reply:
There is some cost, associated with creating a KmsClient instance object, including a potential interaction with a remote KMS Server. Therefore, the cache is a singleton, so only one thread has to pay this cost, others will get the instance from the cache.

Reply:
I'm generally skeptical for forcing singletons into the API.  For instance in Java, I would assume that applications would maintain the an instance of KeyToolkit as singleton (not enforced at the language level).  

Its not clear if this would be an issue, but if this would the use of singleton here work well if the class was being used in a multi-tenant system?

Reply:
Also, this implementation may not be return an actual singleton, as that method is inlinable?

Reply:
Yes, I wholly agree with @emkornfield. Don't force a singleton upon your users. Big no no. It doesn't play well with unit tests, multi-threading, and other cases where you may wish to have more than one singleton. By asking the user to give you the reference, you leave full control of the design of the user application to the user.

Reply:
The cache entries are stored/retrieved by the access tokens, which enables multi-tenancy. Each caller is able to access only the objects cached with its token, using the token as the cache key. [btw, the validity of the token is verified vs the KMS, but this doesn't relate directly to the current discussion].  Also, the technical challenges of using a singleton can be surmounted with a reasonable effort. So I think that the current design is sound.

However, I can understand the desire to use an alternative to singletons where possible; and it indeed should be possible in the case of this particular encryption interface.
I agree it's possible to provide users with an explicit API to create cache instances, and pass them to parquet. But such API would be difficult to explain, and would likely lead to a situation where many users will run without caches, meaning without KMS RPC optimization (each thread will interact with a KMS server, for every key).
So let me propose the following. This encryption interface has a class called PropertiesDrivenCryptoFactory. It's an anchor; one factory instance would typically be created by a user / tenant, and used to create crypto objects for different files / threads. Therefore, we can make the cache a regular, non-singleton, member in this class. It will be shared then across the threads, and will be limited to this tenant. Naturally, this will need to be well documented for the  PropertiesDrivenCryptoFactory class (already a part of the public API of encryption).
What do you think?

Reply:
@ggershinsky I worry that if the cache is a member of `PropertiesDrivenCryptoFactory`, users of API must keep the life time of `PropertiesDrivenCryptoFactory` object after getting `FileEncryptionProperties`/`FileDecryptionProperties`. If they forget that, it will be easy to get a crash in their programs.
I prefer to let users manage the life time of the cache themselves and pass it into `PropertiesDrivenCryptoFactory`.
Any final decision?

Reply:
Eventually I think 2 above ways are identical. The user have to keep an object alive along the time of their program and these both needs to be well documented.

Reply:
>  if the cache is a member of `PropertiesDrivenCryptoFactory`, users of API must keep the life time of `PropertiesDrivenCryptoFactory` object after getting `FileEncryptionProperties`/`FileDecryptionProperties`. If they forget that, it will be easy to get a crash in their programs.

Nope, the lifetime is for an internal use; the entries (cached by the token) are removed after that time, and automatically re-created upon the next call (the token is checked again via the KMS interaction). Users don't need to handle this.


> let users manage the life time of the cache themselves and pass it into PropertiesDrivenCryptoFactory.
> user have to keep an object alive along the time of their program

I'm afraid this virtually guarantees most of the users will run with a default behavior - without caches, and will suffer from performance hits. On the other hand, if the caches are automatically created in the PropertiesDrivenCryptoFactory, users will get performance optimization out of box. Also, this doesn't involve singletons.

> needs to be well documented

Certainly